fw/mmap_buffer.c

```diff
@@ -0,0 +1,329 @@
+	if (atomic_read(&holder->is_freeing))
+		return -ENOENT;
+	if (atomic_read(&holder->dev_is_opened))
+		return -EBUSY;
+
+	atomic_set(&holder->dev_is_opened, 1);
+	filp->private_data = holder;
+
+	return 0;
+}
+
+static int
+dev_file_close(struct inode *ino, struct file *filp)
+{
+	TfwMmapBufferHolder *holder = filp->private_data;
+
+	atomic_set(&holder->dev_is_opened, 0);
+	return 0;
+}
+
+/*
+ * This function handles the mapping of ring buffers into user space. Each
+ * buffer should be mapped by user space with an offset calculated as
+ * full_buffer_size * cpu_num, where full_buffer_size is the size of buffer data
+ * plus TFW_MMAP_BUFFER_DATA_OFFSET; cpu_num is a number of CPU in a row. This
+ * allows determining which CPU's buffer should be mapped based on the offset.
+ */
+static int
+dev_file_mmap(struct file *filp, struct vm_area_struct *vma)
+{
+	TfwMmapBufferHolder *holder = filp->private_data;
+	TfwMmapBuffer *buf = *this_cpu_ptr(holder->buf);
+	unsigned long pfn, size, area_size, user_addr;
+	int cpu_num, cpu_id, i;
+
+	size = vma->vm_end - vma->vm_start;
+
+#define FULL_SIZE TFW_MMAP_BUFFER_FULL_SIZE(buf->size)
+
+	if (size == TFW_MMAP_BUFFER_DATA_OFFSET)
+		area_size = size;
+	else if (size == FULL_SIZE)
+		area_size = FULL_SIZE;
+	else
+		return -EINVAL;
+
+#define NTH_ONLINE_CPU(n) ({		\
+	int cpu, res = -1, i = 0;	\
+	for_each_online_cpu(cpu) {	\
+		if (i == n) {		\
+			res = cpu;	\
+			break;		\
+		}			\
+		++i;			\
+	}				\
+	res;				\
+})
+
+	cpu_num = vma->vm_pgoff / (area_size / PAGE_SIZE);
+	cpu_id = NTH_ONLINE_CPU(cpu_num);
+	if (cpu_id < 0)
+		return -EINVAL;
+
+	buf = *per_cpu_ptr(holder->buf, cpu_id);
+
+	pfn = page_to_pfn(holder->mem[cpu_id].buf_page);
+	if (remap_pfn_range(vma, vma->vm_start, pfn,
+			    TFW_MMAP_BUFFER_DATA_OFFSET, vma->vm_page_prot))
+		return -EAGAIN;
+
+	if (area_size == FULL_SIZE) { /* entire buffer is mapping */
+		pfn = page_to_pfn(holder->mem[cpu_id].data_page);
+		user_addr = vma->vm_start + TFW_MMAP_BUFFER_DATA_OFFSET;
+		for (i = 0; i < 2; ++i) {
+			if (remap_pfn_range(vma, user_addr, pfn, buf->size,
+					    vma->vm_page_prot)) {
+				return -EAGAIN;
+			}
+			user_addr += buf->size;
+		}
+
+		vma->vm_ops = &dev_vm_ops;
+		atomic_set(&buf->is_ready, 1);
+	}
+
+	return 0;
+
+#undef NTH_ONLINE_CPU
+#undef FULL_SIZE
+}
+
+static void
+dev_file_vm_close(struct vm_area_struct *vma)
+{
+	TfwMmapBufferHolder *holder = vma->vm_file->private_data;
+	TfwMmapBuffer *buf = *this_cpu_ptr(holder->buf);
+
+	atomic_set(&buf->is_ready, 0);
+}
+
+TfwMmapBufferHolder *
+tfw_mmap_buffer_create(const char *filename, unsigned int size)
+{
+	TfwMmapBufferHolder *holder;
+	struct page **page_ptr = NULL;
+	unsigned int order, i, page_cnt;
+	int cpu;
+
+	if (size < TFW_MMAP_BUFFER_MIN_SIZE
+	    || size > TFW_MMAP_BUFFER_MAX_SIZE
+	    || !is_power_of_2(size))
+		return NULL;
+
+	if (filename && strlen(filename) >= TFW_MMAP_BUFFER_MAX_NAME_LEN - 1)
+		return NULL;
+
+	if (holders_cnt + 1 > MAX_HOLDERS)
```

Comment:
Is it expected that this function can be called from several threads? If yes holders_cnt should be atomic

Reply:
No, it should be called only at the init. Mayby it need a comment.